File: 1001-2000/1836. Remove Duplicates From an Unsorted Linked List.java
// Medium
// LinkedList
// Time: O(n), Space: O(n)
// https://leetcode.cn/problems/remove-duplicates-from-an-unsorted-linked-list/

import java.util.*;

class Solution {
    public ListNode deleteDuplicatesUnsorted(ListNode head) {
        Map<Integer, Integer> cnt = new HashMap<>();
        
        countFreq(head, cnt);

        return removeDuplicates(head, cnt);
    }

    private void countFreq(ListNode head, Map<Integer, Integer> cnt) {
        if (head == null) return;
        cnt.put(head.val, cnt.getOrDefault(head.val, 0) + 1);
        countFreq(head.next, cnt);
    }

    private ListNode removeDuplicates(ListNode head, Map<Integer, Integer> cnt) {
        if (head == null) return null;

        if (cnt.get(head.val) > 1) {
            // skip
            return removeDuplicates(head.next, cnt);
        } else {
            head.next = removeDuplicates(head.next, cnt);
            return head;
        }
    }
}

File: 1001-2000/1836. Remove Duplicates From an Unsorted Linked List.py
# Time:O(n), Space:O(n)

from typing import Optional, Counter

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def deleteDuplicatesUnsorted(self, head: Optional[ListNode]) -> Optional[ListNode]:
        freq = Counter()
        cur = head
        while cur:
            freq[cur.val] += 1
            cur = cur.next

        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        cur = head

        while cur:
            if freq[cur.val] > 1:
                prev.next = cur.next
            else:
                prev = cur
            cur = cur.next

        return dummy.next
    